### PR TITLE
Dev ali oss persistor

### PR DESCRIPTION
**Pull request by HawkeWoo**
**Proposed changes**:
- [Issue #6237](https://github.com/RasaHQ/rasa/issues/6237): Add Aliyun oss persistor,  this [persistor](https://github.com/HawkeWoo/rasa-alipersistor) support to store the rasa model in aliyun oss. To use this persistor:
`rasa run --remote-storage ali-oss`

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @dakshvar22 will take a look at it as soon as possible ✨

File: rasa/nlu/persistor.py [hunk 1]
```diff
@@ -28,6 +28,20 @@ def get_persistor(name: Text) -> Optional["Persistor"]:
...
+            access_key_id=os.environ.get("ALI_ACCESS_KEY_ID"),
+            access_key_secret=os.environ.get("ALI_ACCESS_KEY_SECRET"),
+            sts_role_arn=os.environ.get("ALI_STS_ROLE_ARN"),
+            region_id=os.environ.get("ALI_REGION_ID"),
+            endpoint=os.environ.get("ALI_ENDPOINT"),
+            bucket_name=os.environ.get("ALI_BUCKET_NAME"),
+            bucket_path=os.environ.get("ALI_BUCKET_PATH"),
+            local_path=os.environ.get("ALI_LOCAL_PATH"),
+        )
+
```
Comment (wochinge) [inline rasa/nlu/persistor.py:44]: @HawkeWoo I thought of a more general solution like the one below. It has the advantage that this also works for other users who provide their own persistors.
```suggestion
        try:
            persistor = rasa.utils.common.class_from_module_path(name)
            return persistor()
        except ImportError:
            raise ImportError(f"Unknown model persistor {name}. Please make sure to "
                              f"either use an included model persistor (`aws`, `gcs` "
                              f"or `azure`) or specify the module path to an external "
                              f"model persistor.")
```

Comment (HawkeWoo) [reply rasa/nlu/persistor.py]: > @HawkeWoo I thought of a more general solution like the one below. It has the advantage that this also works for other users who provide their own persistors.

That sound good, I had change my code and import the persistor by module path.
For example, we can use aliyun oss persistor like:
`rasa run --remote-storage ali_persistor.OssPersistor`
and the external persistor class will use the environment variables: `RASA_PERSISTOR` to init  if nesserary.

File: rasa/nlu/persistor.py [hunk 2]
```diff
@@ -27,15 +30,20 @@ def get_persistor(name: Text) -> Optional["Persistor"]:
...
-                              f"model persistor.")
+            raise ImportError(
+                f"Unknown model persistor {name}. Please make sure to "
+                "either use an included model persistor (`aws`, `gcs` "
+                "or `azure`) or specify the module path to an external "
+                "model persistor."
+            )
+        else:
+            cfg = os.environ.get("RASA_PERSISTOR")
+            kwargs = json.loads(cfg) if cfg else {}
```
Comment (HawkeWoo) [inline rasa/nlu/persistor.py:45]: in aliyun oss persistor case, i will set the environment variables `RASA_PERSISTOR` to  `'{"access_key_id": "*", "access_key_secret": "*", "sts_role_arn": "*", "region_id": "*", "endpoint": *, "bucket_name": "*", "bucket_path": "*", "local_path": "*"}'`

Comment (HawkeWoo) [reply rasa/nlu/persistor.py:45]: > in aliyun oss persistor case, i will set the environment variables `RASA_PERSISTOR` to `'{"access_key_id": "*", "access_key_secret": "*", "sts_role_arn": "*", "region_id": "*", "endpoint": *, "bucket_name": "*", "bucket_path": "*", "local_path": "*"}'`

@wochinge how do you think about this?

Comment (HawkeWoo) [reply rasa/nlu/persistor.py:45]: or i think the more convenient way to use the external persistor is that getting the init parameter from environment on the persistor lib directly

File: tests/nlu/test_persistor.py [hunk 3]
```diff
@@ -165,6 +165,53 @@ def mocked_list_blobs(container_name, prefix=None):
...
+        raise ImportError
+
+    with patch.object(persistor, "get_persistor", mocked_get_persistor):
+        try:
+            _ = persistor.get_persistor("persistor_name")
+        except Exception as e:
+            assert isinstance(e, ImportError)
+
+
 # noinspection PyPep8Naming
```
Comment (HawkeWoo) [inline tests/nlu/test_persistor.py:215]: I have add some tests here

File: docs/docs/user-guide/cloud-storage.mdx [hunk 4]
```diff
@@ -62,6 +62,13 @@ Rasa supports using [S3](https://aws.amazon.com/s3/) ,
 
   If there is no container with the name `AZURE_CONTAINER`, Rasa will create it.
 
+* Other Remote Storage
+
+  Other remote storage is supported using the external model persistor library,
+  which you can install or implement by yourself.
+
+  Start the Rasa server with ``remote-storage`` option set to the module path of model persistor.
```
Comment (wochinge) [inline docs/docs/user-guide/cloud-storage.mdx:70]: I suggest to give an example of how the command would look like

File: docs/docs/user-guide/cloud-storage.mdx [hunk 5]
```diff
@@ -62,6 +62,13 @@ Rasa supports using [S3](https://aws.amazon.com/s3/) ,
 
   If there is no container with the name `AZURE_CONTAINER`, Rasa will create it.
 
+* Other Remote Storage
+
+  Other remote storage is supported using the external model persistor library,
+  which you can install or implement by yourself.
```
Comment (wochinge) [inline docs/docs/user-guide/cloud-storage.mdx:68]:
```suggestion
  Support for other remote storages can be achieved by using an external library or implementing
  a custom module for it by extending the `rasa.nlu.persistor.Persistor` class.
```

File: rasa/nlu/persistor.py [hunk 6]
```diff
@@ -1,18 +1,21 @@
...
+
 
 logger = logging.getLogger(__name__)
 
 
 def get_persistor(name: Text) -> Optional["Persistor"]:
     """Returns an instance of the requested persistor.
 
-    Currently, `aws`, `gcs` and `azure` are supported"""
+    Currently, `aws`, `gcs`, `azure` and 'aliyun oss' are supported"""
```
Comment (wochinge) [inline rasa/nlu/persistor.py:18]:
```suggestion
    Currently, `aws`, `gcs`, `azure` and providing module paths are supported remote storages."""
```

File: rasa/nlu/persistor.py [hunk 7]
```diff
@@ -173,7 +186,7 @@ def _retrieve_tar(self, model_path: Text) -> None:
 class GCSPersistor(Persistor):
     """Store models on Google Cloud Storage.
 
-     Fetches them when needed, instead of storing them on the local disk."""
+    Fetches them when needed, instead of storing them on the local disk."""
```
Comment (wochinge) [inline rasa/nlu/persistor.py:189]: 🙏

File: tests/nlu/test_persistor.py [hunk 8]
```diff
@@ -165,6 +165,53 @@ def mocked_list_blobs(container_name, prefix=None):
...
+
+        def mocked_list_models():
+            return ["model_name"]
+
+        p.list_models = mocked_list_models
+        return p
+
+    with patch.object(persistor, "get_persistor", mocked_get_persistor):
+        result = persistor.get_persistor("persistor_name").list_models()
+    assert result == ["model_name"]
```
Comment (wochinge) [inline tests/nlu/test_persistor.py:180]: an easy test without custom module would be to just the specify the full module path to the `aws` persistor and then check with `isinstance` whether it's the right class. This way you don't have to mock anything.

File: tests/nlu/test_persistor.py [hunk 9]
```diff
@@ -165,6 +165,53 @@ def mocked_list_blobs(container_name, prefix=None):
...
+
+def test_get_external_persistor_raise_exeception():
+    def mocked_get_persistor(name):
+        raise ImportError
+
+    with patch.object(persistor, "get_persistor", mocked_get_persistor):
+        try:
+            _ = persistor.get_persistor("persistor_name")
+        except Exception as e:
+            assert isinstance(e, ImportError)
```
Comment (wochinge) [inline tests/nlu/test_persistor.py:212]: I'd mock `class_from_module_path` so we can actually see if `get_persistor` handles it correctly.
There is also a function `pytest.raises` which makes checking for exceptions very handy.

File: docs/docs/user-guide/cloud-storage.mdx (hunk 4 above)
Comment (HawkeWoo) [reply docs/docs/user-guide/cloud-storage.mdx:70]: I had add a single command

File: tests/nlu/test_persistor.py [hunk 10]
```diff
@@ -165,51 +165,9 @@ def mocked_list_blobs(container_name, prefix=None):
...
-        raise ImportError
-
-    with patch.object(persistor, "get_persistor", mocked_get_persistor):
-        try:
-            _ = persistor.get_persistor("persistor_name")
-        except Exception as e:
-            assert isinstance(e, ImportError)
+def test_get_external_persistor():
+    p = persistor.get_persistor("rasa.nlu.persistor.Persistor")
+    assert isinstance(p, persistor.Persistor)
```
Comment (HawkeWoo) [inline tests/nlu/test_persistor.py:170]: @wochinge  It seems that only one test is enough -.-

File: docs/docs/user-guide/cloud-storage.mdx [hunk 11]
```diff
@@ -62,6 +62,19 @@ Rasa supports using [S3](https://aws.amazon.com/s3/) ,
 
   If there is no container with the name `AZURE_CONTAINER`, Rasa will create it.
 
+* Other Remote Storage
+
+  Support for other remote storages can be achieved by using an external library or implementing
+  a custom module for it by extending the `rasa.nlu.persistor.Persistor` class.
+
+  Start the Rasa server with ``remote-storage`` option set to the module path of model persistor.
```
Comment (wochinge) [inline docs/docs/user-guide/cloud-storage.mdx:70]:
```suggestion
  Start the Rasa server with `remote-storage` option set to the module path of model persistor.
```

File: docs/docs/user-guide/cloud-storage.mdx [hunk 12]
```diff
@@ -62,6 +62,19 @@ Rasa supports using [S3](https://aws.amazon.com/s3/) ,
...
+
+  A single command is:
```
Comment (wochinge) [inline docs/docs/user-guide/cloud-storage.mdx:72]:
```suggestion
  An example command is:
```

File: docs/docs/user-guide/cloud-storage.mdx [hunk 13]
```diff
@@ -62,6 +62,19 @@ Rasa supports using [S3](https://aws.amazon.com/s3/) ,
...
+
+  ```bash
+  rasa run --remote-storage rasa.nlu.persistor.Persistor
```
Comment (wochinge) [inline docs/docs/user-guide/cloud-storage.mdx:75]: That should be a custom module. People might misunderstand if we give the parent class as example here.
```suggestion
  rasa run --remote-storage <your module>.<class name>
```

File: tests/nlu/test_persistor.py (hunk 10 above)
Comment (wochinge) [reply tests/nlu/test_persistor.py:170]: can we please also add a test for an invalid module? Otherwise we're ready to go 🎉

**Comment by HawkeWoo** [comment]
> Looks great! Can you please also add a changelog?

sure, and i also added a test for an invalid persistor module

**Pull request merged**